### PR TITLE
DOC: Document design principles.

### PR DESCRIPTION
**Pull request by danielballan**
The purpose of this section is the provide public documentation hashing out
arguments about design that we find ourselves making often in person.

### CODE REVIEW

File: docs/source/guiding-design-principles.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+:func:`numpy.sum` even though pandas was created long after :func:`numpy.sum`.
+This is because :func:`numpy.sum` avoids assuming it will be passed specific
+data types; it accepts any object that provides the right methods (interfaces).
+Where possible, avoid ``isinstance`` checks in your code, and try to make your
+functions work on the broadest possible range of input types.
+
+"Stop Writing Classes"
+----------------------
+
+It is often tempting to invent special objects for a use case or workflow ---
```
Comment (jrmlhermitte) [inline docs/source/guiding-design-principles.rst]: :+1:

**Comment by danielballan** [comment]
@jrmlhermitte Thanks, that's a solid idea. I have added a commit with your code (basically as written) with a little introduction and conclusion. What do you think?

File: docs/source/guiding-design-principles.rst [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    def get_image(filename, options={}):
+        ...
+
+Although the complexity appears to have been reduced through hidden keyword
+arguments, it has been slightly complicated through the need to dig through
+more documentation to better understand how to use them.
+
+Because new science occurs when old ideas are reapplied or extended in
+unforeseen ways, scientific code should should not bury its complexity and
+overly optimize for a specific use case. It should expose what complexity there
```
Comment (jrmlhermitte) [inline docs/source/guiding-design-principles.rst]: :+1:

**Comment by jrmlhermitte** [comment]
I like it! Looks good to me thanks!

File: docs/source/guiding-design-principles.rst [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+data structures and functions.
+
+As another example, the widely-used scikit-image library initially experimented
+with using an ``Image`` class, but ultimately decided that it was better to use
+plain old numpy arrays. All scientific Python libraries understand numpy
+arrays, but they don't understand custom classes, so it is better to pass
+application-specific metadata *alongside* a standard array than to try to
+encapsulate all of that information in a new, bespoke object.
+
+Complexity is Always Conserved
```
Comment (tacaswell) [inline docs/source/guiding-design-principles.rst]: and strictly greater than the system you are modeling.

**Pull request merged**